src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetCommandHandler.cs

```diff
@@ -100,6 +102,23 @@ public CommandState GetCommandState(TabKeyCommandArgs args)
             return CommandState.Available;
         }
 
+        public CommandState GetCommandState(AutomaticLineEnderCommandArgs args, Func<CommandState> nextCommandHandler)
+        {
+            return nextCommandHandler();
+        }
+
+        public void ExecuteCommand(AutomaticLineEnderCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            AssertIsForeground();
+            if (AreSnippetsEnabled(args)
+                && args.TextView.Properties.TryGetProperty(typeof(AbstractSnippetExpansionClient), out AbstractSnippetExpansionClient snippetExpansionClient))
+            {
+                snippetExpansionClient.TryHandleReturn();
+            }
+
+            nextCommandHandler();
+        }
+
         public bool ExecuteCommand(ReturnKeyCommandArgs args, CommandExecutionContext context)
         {
             AssertIsForeground();
```

Comment:
aside.  this method is a monstrosity.

Reply:
![image](https://user-images.githubusercontent.com/1408396/102509112-02411580-403b-11eb-9327-57a37ab6b527.png)